Content.Server/Internals/InternalsProvider.cs

```diff
@@ -0,0 +1,161 @@
+            if (args.ForcedState == true || (args.ForcedState == null && !component.Connected))
+                ConnectToInternals(uid, component);
+            else if (args.ForcedState == false || (args.ForcedState == null && component.Connected))
+                DisconnectFromInternals(uid, component);
+
+            args.Handled = wasConnected != component.Connected || component.Connected == args.ForcedState;
+
+            if (args.Actions is not null)
+                args.Actions.GrantOrUpdate(ItemActionType.ToggleInternals, GetInternalsComponent(component) != null, component.Connected);
+            else
+            {
+                ItemActionsComponent? actions = null;
+                if (Resolve(uid, ref actions))
+                    actions.GrantOrUpdate(ItemActionType.ToggleInternals, GetInternalsComponent(component) != null, component.Connected);
+            }
+
+            if (args.BypassCheck && !args.Handled)
+                Logger.ErrorS("internals", $"Internals @{uid} state change failed to change to \"{args.ForcedState}\" despite being forced to bypass a usability check");
+        }
+
+        public void ConnectToInternals(EntityUid uid, InternalsProviderComponent component)
+        {
+            if (component.Connected) return;
+
+            var internals = GetInternalsComponent(component);
+            if (internals?.TryConnectTank(component.Owner) ?? false)
+                component.Internals = internals;
+        }
+
+        public void DisconnectFromInternals(EntityUid uid, InternalsProviderComponent component)
+        {
+            if (!component.Connected) return;
+
+            //Logger.DebugS("internals", $"Disconnecting internals @{uid} from {component.Internals!.Owner.Name}");
+
+            var internals = component.Internals!;
+            component.Internals = null;
+            internals.DisconnectTank(true);
+
+            ItemActionsComponent? actions = null;
+            if (Resolve(uid, ref actions))
+                actions.GrantOrUpdate(ItemActionType.ToggleInternals, GetInternalsComponent(component) != null, component.Connected);
+        }
+
+        // It's rather hacky but it only needs to happen once, when trying to turn it on, so it should be alright?
+        // Kinda a fair bit of edge cases here
+        private InternalsComponent? GetInternalsComponent(InternalsProviderComponent component)
+        {
+            var owner = component.Owner;
+            if (owner.Deleted) return null;
+
+            if (!owner.TryGetContainer(out var container))
+                return null;
+
+            if (container.Owner.TryGetComponent(out InternalsComponent? internals))
+                return internals!;
+
+            if (container.Owner.TryGetComponent(out InternalsSuitComponent? suit))
+                if (suit.Owner.TryGetContainer(out var suitContainer))
+                    return suitContainer?.Owner.GetComponentOrNull<InternalsComponent>();
+
+            return null;
```

Comment:
Use the EntityManager methods and avoid using the IEntity ones (we'd obsolete but there's 4,000 callers).